samples/apps/logging/logging.cpp

```diff
@@ -1486,6 +1530,41 @@ namespace loggingapp
         {ccf::user_signature_auth_policy})
         .set_auto_schema<void, std::string>()
         .install();
+
+      auto fresh_scope = [this](auto& ctx, nlohmann::json&&) {
+        auto prefix = get_scope(ctx);
+
+        std::string scope = prefix.value_or("");
+
+        constexpr size_t max_tries = SIZE_MAX;
+        for (size_t i = 0; i < max_tries; i++)
+        {
+          size_t pub_sz =
+            ctx.tx.template ro<RecordsMap>(public_records({scope}))->size();
+          size_t prv_sz =
+            ctx.tx.template ro<RecordsMap>(private_records({scope}))->size();
+          if (pub_sz + prv_sz == 0)
+          {
+            return ccf::make_success(LoggingFreshScope::Out{scope});
+          }
+          scope = prefix.value_or("") + "_" + std::to_string(i + 1);
+        }
+
+        return ccf::make_error(
+          HTTP_STATUS_INTERNAL_SERVER_ERROR,
+          ccf::errors::InternalError,
+          fmt::format(
+            "Could not find a suitable scope after {} tries", max_tries));
+      };
+
+      make_read_only_endpoint(
+        "/log/fresh_scope",
+        HTTP_GET,
+        ccf::json_read_only_adapter(fresh_scope),
+        auth_policies)
+        .add_query_parameter<std::string>("scope")
+        .set_auto_schema<void, LoggingFreshScope::Out>()
+        .install();
```

Comment:
Since we don't record the fresh scope in the KV, it's possible to re-use it if we don't have a `POST` in between:

```
scope1 = GET /app/log/fresh_scope?scope=foo # scope1 == foo_1
scope2 = GET /app/log/fresh_scope?scope=foo # scope2 == foo_2

# Now these conflict, despite using fresh_scope!
POST /app/log/private?scope={scope1}
POST /app/log/private?scope={scope2}
```

I think this is a bad pattern to include in the sample app. We should either add a new table to store the last `i` assigned to a fresh scope (so `fresh_scope` includes a read and write, and reliably gives each caller a unique response), or roll it into the first `POST` request (so we find the first table with no writes, then write to it _and_ tell you in the response how to write to this in future). Alternatively, if possible it'd be preferable to make the scopes entirely client determined - the caller is responsible for finding a unique string name for their scope, and the server doesn't help order multiple callers if they're using the same name.

Reply:
Hmm. TBH, I think this is fine. A scope that ends up not being used can be re-used. If the client wants to make sure that a particular scope is reserved to it, it would have to write something to it. Like I mentioned further above, we could move this completely into Python scripts and just check for table size == 0 when we require a fresh scope (and we can keep scoped counters to get the next one immediately too).